src/geqrf_qdwh_full.cc

```diff
@@ -0,0 +1,490 @@
+// Copyright (c) 2017-2022, University of Tennessee. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+// This program is free software: you can redistribute it and/or modify it under
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#include "slate/slate.hh"
+#include "auxiliary/Debug.hh"
+#include "slate/Matrix.hh"
+#include "internal/internal.hh"
+
+namespace slate {
+
+// specialization namespace differentiates, e.g.,
+// internal::geqrf_qdwh_full from internal::specialization::geqrf
+namespace internal {
+namespace specialization {
+
+//------------------------------------------------------------------------------
+/// An auxiliary routine to find each rank's first (top-most) row
+/// in panel k.
+///
+/// @param[in] A_panel
+///     Current panel, which is a sub of the input matrix $A$.
+///
+/// @param[in] k
+///     Index of the current panel in the input matrix $A$.
+///
+/// @param[out] first_indices
+///     The array of computed indices.
+///
+/// @ingroup geqrf_qdwh_full_specialization
+///
+template <typename scalar_t>
+void geqrf_compute_first_indices(Matrix<scalar_t>& A_panel, int64_t k,
+                                 std::vector< int64_t >& first_indices)
```

Comment:
This routine already exists in geqrf.cc, right? Hopefully with identical implementation. We shouldn't define the same name in 2 different files. Because of templates, it will compile fine, but it's up to the compiler which implementation gets used. It should be moved to a header file. (This happened for some other function, I think with Neil. We can ask what his solution was.)

Reply:
Right, they are identical. what about have it in internal_householder.hh?

Reply:
I moved it to internal_util.hh as suggested.